bpf/flows.c

```diff
@@ -95,59 +151,137 @@ static inline int fill_ip6hdr(struct ipv6hdr *ip, void *data_end, struct flow *f
     return SUBMIT;
 }
 // sets flow fields from Ethernet header information
-static inline int fill_ethhdr(struct ethhdr *eth, void *data_end, struct flow *flow) {
+static inline int fill_ethhdr(struct ethhdr *eth, void *data_end, flow_id_v *flow_id, u32 *flags) {
     if ((void *)eth + sizeof(*eth) > data_end) {
         return DISCARD;
     }
-    __builtin_memcpy(flow->data_link.dst_mac, eth->h_dest, ETH_ALEN);
-    __builtin_memcpy(flow->data_link.src_mac, eth->h_source, ETH_ALEN);
-    flow->protocol = __bpf_ntohs(eth->h_proto);
-    // TODO: ETH_P_IPV6
-    if (flow->protocol == ETH_P_IP) {
+    __builtin_memcpy(flow_id->dst_mac, eth->h_dest, ETH_ALEN);
+    __builtin_memcpy(flow_id->src_mac, eth->h_source, ETH_ALEN);
+    flow_id->eth_protocol = __bpf_ntohs(eth->h_proto);
+
+    if (flow_id->eth_protocol == ETH_P_IP) {
         struct iphdr *ip = (void *)eth + sizeof(*eth);
-        return fill_iphdr(ip, data_end, flow);
-    } else if (flow->protocol == ETH_P_IPV6) {
+        return fill_iphdr(ip, data_end, flow_id, flags);
+    } else if (flow_id->eth_protocol == ETH_P_IPV6) {
         struct ipv6hdr *ip6 = (void *)eth + sizeof(*eth);
-        return fill_ip6hdr(ip6, data_end, flow);
+        return fill_ip6hdr(ip6, data_end, flow_id, flags);
+    } else {
+        // TODO : Need to implement other specific ethertypes if needed
+        // For now other parts of flow id remain zero
+        memset (&(flow_id->src_ip),0, sizeof(struct in6_addr));
+        memset (&(flow_id->dst_ip),0, sizeof(struct in6_addr));
+        flow_id->protocol = 0;
```

Comment:
just to understand: this is a corner case where eth_proto is neither ipv4 nor ipv6 .. in that case we don't fill ip/proto information, however `flow_id->eth_protocol` is still present and exported in flows, right?

Reply:
Yes, exactly.